File: __iOS_BACKUP__/ShapeByte/shared/ui/dimensions/tests/DimensionProviderTest.swift
//
//  DimensionProviderTest.swift
//  ShapeByteTests
//
//  Created by Lang, Stefan [RTL Tech] on 14.08.24.
//

import XCTest

final class DimensionProviderTest: XCTestCase {

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testWithDimensionalAspectHeight() throws {
        // screen and default size are the same, return same height
        var sut = createSUT(
            defaultSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
            deviceScreenSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size
        )

        let height: CGFloat = 40.0
        var result = sut.withDimensionalAspect(height: height)
        XCTAssertEqual(height, result)

        // screen size larger than default, height should be larger
        sut = createSUT(
            defaultSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
            deviceScreenSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.large.size
        )

        result = sut.withDimensionalAspect(height: height)
        XCTAssertEqual(44.170616112, result, accuracy: 0.2)

        // screen size smaller than default, height should be smaller
        sut = createSUT(
            defaultSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
            deviceScreenSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.small.size
        )

        result = sut.withDimensionalAspect(height: height)
        XCTAssertEqual(31.611374408, result, accuracy: 0.2)
    }

    func testWithDimensionalAspectHeightWithMax() throws {
        // screen size larger than default, but height should be same since max is
        var sut = createSUT(
            defaultSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
            deviceScreenSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.large.size
        )

        let height: CGFloat = 40.0
        var result = sut.withDimensionalAspect(height: height, max: height)
        XCTAssertEqual(height, result)

        // screen size smaller than default, height should be smaller, but max is even smaller
        sut = createSUT(
            defaultSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
            deviceScreenSize: DeviceSizeCategoryProvider.iPhoneSizeBounds.small.size
        )

        let max = 20.0
        result = sut.withDimensionalAspect(height: height, max: max)
        XCTAssertEqual(max, result)
    }

    private func createSUT(
        defaultSize: CGSize = DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
        deviceScreenSize: CGSize = DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size
    ) -> DimensionProvider {
        let sut = DimensionProvider(
            defaultSize: defaultSize,
            deviceSizeCategoryProvider: DeviceSizeCategoryProvider()
                .setup(
                    screenSize: deviceScreenSize
                )
        )

        return sut
    }

}

File: foundation/ui/api/src/commonTest/kotlin/de/stefan/lang/foundationUi/api/dimension/DimensionProviderTest.kt
package de.stefan.lang.foundationUi.api.dimension

import de.stefan.lang.foundationCore.api.devicesize.DeviceSizeCategory
import de.stefan.lang.foundationCore.api.devicesize.Size
import de.stefan.lang.foundationCore.test.devicesize.DeviceSizeCategoryProviderMock
import de.stefan.lang.foundationCore.test.devicesize.ScreenSizeProviderMock
import kotlin.test.Test
import kotlin.test.assertEquals

class DimensionProviderTest {
    private val sizeCategories = listOf(
        DeviceSizeCategory.XSmall(size = Size(width = 240f, height = 480f)),
        DeviceSizeCategory.Small(size = Size(width = 360f, height = 640f)),
        DeviceSizeCategory.Medium(size = Size(width = 480f, height = 800f)),
        DeviceSizeCategory.Large(size = Size(width = 560f, height = 900f)),
    )

    private val testDeviceCategory = sizeCategories
        .filterIsInstance<DeviceSizeCategory.Medium>()
        .first()

    private val xSmallDeviceCategory = sizeCategories
        .filterIsInstance<DeviceSizeCategory.XSmall>()
        .first()

    @Test
    fun `should return smaller dimension`() {
        val screenSize = xSmallDeviceCategory.size
        val sut = createSUT(screenSize, xSmallDeviceCategory)

        val dimension = 84f
        val result = sut.withDimensionalAspect(dimension)
        assertEquals(result, 50.4f)
    }

    @Test
    fun `should return dimension if max not set`() {
        val screenSize = testDeviceCategory.size
        val sut = createSUT(screenSize, testDeviceCategory)

        val dimension = 84f
        val result = sut.withDimensionalAspect(dimension)
        assertEquals(result, dimension)
    }

    @Test
    fun `should return dimension if max is greater`() {
        val screenSize = testDeviceCategory.size
        val sut = createSUT(screenSize, testDeviceCategory)

        val dimension = 84f
        val result = sut.withDimensionalAspect(dimension, dimension * 2)
        assertEquals(result, dimension)
    }

    @Test
    fun `should return max if dimension is greater`() {
        val screenSize = Size(width = 400.0f, height = 800.0f)
        val sut = createSUT(screenSize, testDeviceCategory)

        val dimension = 84f
        val max = dimension / 2
        val result = sut.withDimensionalAspect(dimension, max)
        assertEquals(result, max)
    }

    private fun createSUT(
        screenSize: Size,
        deviceSizeCategory: DeviceSizeCategory,
    ): DimensionProvider {
        val deviceSizeCategoryProvider = DeviceSizeCategoryProviderMock(
            ScreenSizeProviderMock(screenSize),
            deviceSizeCategory,
            sizeCategories,
        )

        return DimensionProvider(deviceSizeCategoryProvider)
    }
}
